src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingReadStream.cs

```diff
@@ -0,0 +1,244 @@
+        private readonly Decoder _decoder;
+        private readonly Encoder _encoder;
+
+        private ArraySegment<byte> _byteBuffer;
+        private ArraySegment<char> _charBuffer;
+        private ArraySegment<byte> _overflowBuffer;
+        private bool _disposed;
+
+        public TranscodingReadStream(Stream input, Encoding sourceEncoding)
+        {
+            _stream = input;
+
+            // The "count" in the buffer is the size of any content from a previous read.
+            // Initialize them to 0 since nothing has been read so far.
+            _byteBuffer = new ArraySegment<byte>(ArrayPool<byte>.Shared.Rent(MaxByteBufferSize), 0, count: 0);
+
+            // Attempt to allocate a char buffer than can tolerate the worst-case scenario for this
+            // encoding. This would allow the byte -> char conversion to complete in a single call.
+            // However limit the buffer size to prevent an encoding that has a very poor worst-case scenario.
+            // The conversion process is tolerant of char buffer that is not large enough to convert all the bytes at once.
+            int maxCharBufferSize = Math.Min(MaxCharBufferSize, sourceEncoding.GetMaxCharCount(MaxByteBufferSize));
+            _charBuffer = new ArraySegment<char>(ArrayPool<char>.Shared.Rent(maxCharBufferSize), 0, count: 0);
+
+            _overflowBuffer = new ArraySegment<byte>(ArrayPool<byte>.Shared.Rent(OverflowBufferSize), 0, count: 0);
+
+            _decoder = sourceEncoding.GetDecoder();
+            _encoder = Encoding.UTF8.GetEncoder();
+        }
+
+        public override bool CanRead => true;
+        public override bool CanSeek => false;
+        public override bool CanWrite => false;
+        public override long Length => throw new NotSupportedException();
+
+        public override long Position
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        internal int ByteBufferCount => _byteBuffer.Count;
+        internal int CharBufferCount => _charBuffer.Count;
+        internal int OverflowCount => _overflowBuffer.Count;
+
+        public override void Flush() { }
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (offset < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            if (buffer.Length - offset < count)
+            {
+                throw new ArgumentOutOfRangeException();
```

Comment:
We generally use ArgumentException for this, with a message that conveys what went wrong, e.g. https://github.com/dotnet/runtime/blob/ccada164f3a18207e91f1e3d0dc7c168ae3de0d4/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs#L341